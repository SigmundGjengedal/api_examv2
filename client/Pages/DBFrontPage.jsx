import {Link, useNavigate} from "react-router-dom";
import React from "react";

export function DBFrontPage() {
    const navigate = useNavigate();
    async function handleLogout() {
        await fetch("/api/login", { method: "delete" });
        navigate("/")
    }
    return (
        <div>
            <h1>Movie Database</h1>
            <ul>
                <div>
                    <Link to={"/movies"}>List movies</Link>
                </div>
                <div>
                    <Link to={"/movies/new"}>Add new movie</Link>
                </div>
                <div>
                    <Link to="/profile">Your Profile</Link>
                </div>
                <div>
                    <Link to="/chat">Chat</Link>
                </div>
                <div>
                    <button onClick={handleLogout}>Log out</button>
                </div>
            </ul>
        </div>
    );
}