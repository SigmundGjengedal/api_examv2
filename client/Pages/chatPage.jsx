import React, {useState} from "react";
import "../styling/chatPage.css"

function Login({onLoginCallback}) {
    const [username, setUsername] = useState("");

    function handleSubmit(event) {
        event.preventDefault()
        onLoginCallback(username) // sender username tilbake til ChatPage.
    }

    // form der man angir brukernavn. Det lagres i state.
    return <div>
        <h1>please login to chat</h1>
        <form onSubmit={handleSubmit}>
            <label>
                Username:
                <input
                    value={username}
                    onChange={(e) => setUsername(e.target.value)}
                />
            </label>
            <button>Login</button>
        </form>
    </div>
}

function ChatApp({username}) {
    function handleMessage(event){
        event.preventDefault()
    }
    return (
        <div className={"chat-page"}>
            <header> Chat with {username}</header>
            <main> Here is the main content</main>
            <footer>
                <form onSubmit={handleMessage}>
                    <input />
                    <button>Send</button>
                </form>
            </footer>
        </div>
    )
}

export function ChatPage() {
    const [username, setUsername] = useState("Sigmund");

    // går hit å henter username.
    if (!username) {
        return <Login onLoginCallback={(username) => setUsername(username)}/>
    }
    return <ChatApp username={username}/>
}