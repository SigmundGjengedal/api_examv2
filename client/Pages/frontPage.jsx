import {Link} from "react-router-dom";
import React from "react";

export function FrontPage() {
    return (
        <>
            <div>
                <h1 style={{padding: "20px"}}>Front Page</h1>
            </div>
            <div className="flex-container">
                <div className="column">
                    <Link to={"/movies"}>List movies</Link>
                </div>
                <div className="column">
                    <Link to="/login">Login</Link>
                </div>
            </div>
        </>
    );
}