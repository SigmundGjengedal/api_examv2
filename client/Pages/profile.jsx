import {useLoading} from "../customHooks/useLoading";
import {fetchJSON} from "../http";
import React from "react";

export function Profile() {
    const {loading, data, error} = useLoading(async () => {
        return await fetchJSON("/api/login");
    });
    console.log(data)

    if (loading) {
        return <div>Please wait...</div>;
    }
    if (error) {
        return <div>Error! {error.toString()}</div>;
    }
    return (
        <div>
            <h1>User Profile</h1>
            <div>
                {data.user.google && (
                    <div>
                        <h2>Logged in as {data.user.google.name}({data.email})</h2>
                        <img src={data.user.google.picture} alt={"Profile picture"}/>
                    </div>
                )}
                {data.user.hk && (
                    <div>
                        <h2>Logged in as {data.user.hk.name}({data.email})</h2>
                        <img src={data.user.hk.picture} alt={"No Profile picture"}/>
                    </div>
                )}

            </div>
            <div>

            </div>
        </div>
    );
}