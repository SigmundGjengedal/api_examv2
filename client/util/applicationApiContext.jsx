import React from "react";
import {fetchJSON} from "../http";
import {postJSON} from "../http";

export const ApplicationApiContext = React.createContext({
    async listMovies(query) {
        return await fetchJSON("/api/movies?" + new URLSearchParams(query));
    },
    async fetchLogin() {
        return await fetchJSON("/api/login");
    },
    async registerLogin(provider, login) {
        return await postJSON(`/api/login/${provider}`, login);
    },
    async createMovie(movie) {
        fetch("/api/movies/new", {
            method: "post",
            body: JSON.stringify(movie),
            headers: {
                "Content-Type": "application/json",
            },
        });
    },
});