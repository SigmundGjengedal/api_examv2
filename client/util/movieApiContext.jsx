import React from "react";
import {fetchJSON} from "../http";

export const MovieApiContext = React.createContext({
    async listMovies(query) {
        return await fetchJSON("/api/movies?" + new URLSearchParams(query));
    },
    async createMovie(movie) {
        fetch("/api/movies/new", {
            method: "post",
            body: JSON.stringify(movie),
            headers: {
                "Content-Type": "application/json",
            },
        });
    },
});