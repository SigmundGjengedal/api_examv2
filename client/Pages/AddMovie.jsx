import React, {useContext, useState} from "react";
import {useNavigate} from "react-router-dom";
import {MovieApiContext} from "../util/movieApiContext";

// hm AddMovie
function FormInput({ label, value, onChangeValue }) {
    return (
        <div className="form-input">
            <label>
                <strong>{label}</strong>
                <input value={value} onChange={(e) => onChangeValue(e.target.value)} />
            </label>
        </div>
    );
}

export function AddMovie() {
   // henter createMovie metoden fra MovieApiContext
    const { createMovie } = useContext(MovieApiContext);
    // setter state for form input
    const [title, setTitle] = useState("");
    const [plot, setPlot] = useState("");
    const [year1, setYear] = useState("");
    const [country, setCountry] = useState("");
    const [director, setDirector] = useState("");
    const navigate = useNavigate();

    async function handleSubmit(e) {
        e.preventDefault();
        createMovie({title, year1, plot, country,director})

        setTitle("");
        setYear("");
        setPlot("");
        setCountry("");
        setDirector("");
        navigate("/db-front-page");
    }

    return (
        <form onSubmit={handleSubmit}>
            <h1>Add Movie</h1>
            <FormInput label={"Title:"} value={title} onChangeValue={setTitle} />
            <FormInput label={"Year:"} value={year1} onChangeValue={setYear} />
            <FormInput label={"Country:"} value={country} onChangeValue={setCountry} />
            <FormInput label={"Plot:"} value={plot} onChangeValue={setPlot} />
            <FormInput label={"Director:"} value={director} onChangeValue={setDirector} />
            <button>Submit</button>
        </form>
    );
}