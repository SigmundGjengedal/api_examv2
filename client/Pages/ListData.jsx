import {useLoading} from "../customHooks/useLoading";
import React, {useContext, useState} from "react";
import {MovieApiContext} from "../util/movieApiContext";

export function ListData() {
    const {listMovies} = useContext(MovieApiContext);
    //state for searchbar
    const [country, setCountry] = useState("");
    const [countryQuery, setCountryQuery] = useState("");

    const {loading, error, data} = useLoading(async () => listMovies({country})
        , [country]);

    function handleSubmitQuery(e) {
        e.preventDefault();
        setCountry(countryQuery);
    }

    if (loading) {
        return <div>Loading...</div>;
    }
    if (error) {
        return (
            <div>
                <h1>Error</h1>
                <div id="error-text">{error.toString()}</div>
            </div>
        );
    }

    return (
        <div>
            <h1>Movies in the DataBase</h1>

            <div>
                <form onSubmit={handleSubmitQuery}>
                    <label>
                        Country:
                        <input
                            id="country-query"
                            value={countryQuery}
                            onChange={(e) => setCountryQuery(e.target.value)}
                        />
                        <button>Filter</button>
                    </label>
                </form>
            </div>

            {data.map((movie, index) => (
                <MovieCard key={index} movie={movie}/>
            ))}
        </div>
    );
}

function MovieCard({movie}) {
    return <div>
        <h3>Title: {movie.title}</h3>
        <p>Year: {movie.year}</p>
        <p>Plot: {movie.plot}</p>
        <p>Directors: {movie.directors}</p>
        {
            movie.countries.map((c, index) => {
                return (<p key={index}>Country: {c}</p>)
            })}
        {movie.poster ? <img src={movie.poster} alt="pic" width={100}/> : null}
    </div>;
}