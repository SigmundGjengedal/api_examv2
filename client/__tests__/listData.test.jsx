import {ListData} from "../Pages/ListData";
import {ApplicationApiContext} from "../util/applicationApiContext";
import React from "react";
import ReactDOM from "react-dom";
import {act, Simulate} from "react-dom/test-utils";

// tester for front-end. Ser hvordan data vises i nettleser.
// tester med dummy data. DB testes ikke.
const movies = [
    { title: "movie 1", countries: ["Norway"] },
    { title: "movie 2", countries: ["Norway"] }
];
// hm
async function renderListMovies(listMovies) {
    const element = document.createElement("div");
    await act(async () =>
        ReactDOM.render(
            <ApplicationApiContext.Provider value={{ listMovies }}>
                <ListData />
            </ApplicationApiContext.Provider>,
            element
        )
    );
    return element;
}
describe("ListData component", () => {

    it("shows loading screen", async () => {
        const element = await renderListMovies(() => new Promise(() => {}));
        expect(element.innerHTML).toMatchSnapshot();
    });

    // h3 er fra MovieCard i ListData. Tester med to filmer
    it("shows movies", async () => {
        const element = await renderListMovies(async () => movies)
        expect(
            Array.from(element.querySelectorAll("h3")).map((e) => e.innerHTML)
        ).toEqual(["Title: movie 1", "Title: movie 2"]);
        expect(element.innerHTML).toMatchSnapshot();
    });

    it("queries by search filter (country)", async () => {
        const domElement = document.createElement("div");
        const listMovies = jest.fn(() => []);
        await act(async () => {
            ReactDOM.render(
                <ApplicationApiContext.Provider value={{ listMovies }}> // jest.fn
                    <ListData />
                </ApplicationApiContext.Provider>,
                domElement
            );
        });
        Simulate.change(domElement.querySelector("#country-query"), {
            target: { value: "Norway" },
        });
        await act(async () => {
            await Simulate.submit(domElement.querySelector("form"));
        });
        expect(listMovies).toHaveBeenCalledWith({
            country: "Norway",
        });
    });

    it("shows error in error.toString()", async () => {
        const element = await renderListMovies(async () => {
            throw new Error("Something went wrong");
        });

        expect(element.querySelector("#error-text").innerHTML).toEqual( "Error: Something went wrong");
        expect(element.innerHTML).toMatchSnapshot();
    });

});
