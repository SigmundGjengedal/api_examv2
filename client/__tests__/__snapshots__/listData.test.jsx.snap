// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListData component shows error in error.toString() 1`] = `"<div><h1>Error</h1><div id=\\"error-text\\">Error: Something went wrong</div></div>"`;

exports[`ListData component shows loading screen 1`] = `"<div>Loading...</div>"`;

exports[`ListData component shows movies 1`] = `"<div><h1>Movies in the DataBase</h1><div><form><label>Country:<input id=\\"country-query\\" value=\\"\\"><button>Filter</button></label></form></div><div><h3>Title: movie 1</h3><p>Year: </p><p>Plot: </p><p>Directors: </p><p>Country: Norway</p></div><div><h3>Title: movie 2</h3><p>Year: </p><p>Plot: </p><p>Directors: </p><p>Country: Norway</p></div></div>"`;
